src/test/ui/fmt/format-args-capture-issue-93378.stderr

```diff
@@ -1,19 +1,14 @@
-error: invalid reference to positional arguments 1 and 2 (there is 1 argument)
-  --> $DIR/format-args-capture-issue-93378.rs:5:26
+error: 3 positional arguments in format string, but there is 1 argument
+  --> $DIR/format-args-capture-issue-93378.rs:5:23
    |
 LL |     println!("{a} {b} {} {} {c} {}", c = "c");
-   |                          ^^     ^^
-   |
-   = note: positional arguments are zero-based
+   |                       ^^ ^^     ^^       ---
```

Comment:
```suggestion
   |                       ^^ ^^     ^^       --- only 1 named formatting argument available
   |                       |
   |                       3 positional arguments in the format string
```

Reply:
That seems like something for a separate PR. This refactor is meant as a starting point to make it much easier to make changes like that afterwards, since now the part that does the diagnostics is completely independent of the representation of the expanded `fmt::Arguments`.

For now, I tried to keep the diagnostics mostly unchanged, except for a few bug fixes, and a few cases were it now picks the rignt alternative because the right information is now available. (E.g. when to pick the simple "{N} positional arguments in format string, but .." diagnostic instead of the more general "invalid refernce to positional argument {X}" one.)

Reply:
Fair enough. I started looking at the diagnostics first to get a sense for what might have changed, which has the annoying side effect of me noticing all of these things I'd like to clean up regardless 😅